_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,185 @@
+
+## You're in my language and you'll follow my language's abstract machine's rule!
+
+So what's the point of CHERI if an ISA can't give you these properties?
+CHERI is not designed to give you an object model that you must conform to, it's designed to *give language implementers tools to enforce these properties*.
+
+That's a very important distinction and it becomes even more important when you ask the follow-on question: Against whom, exactly, are these properties being enforced?
+
+We assume that a compiler has some notion of an object model.
+It may be able to enforce that memory model entirely because the source language has strong typing guarantees.
+For example, we have a JavaScript VM in the CHERIoT RTOS repository that builds its own garbage-collected type-safe object model using 16-bit integers to represent object pointers (more on that later).
+That compiler also needs to interoperate with other code.
+The other code may be compiled with the same compiler but [rely on exploiting compiler bugs to bypass certain properties](https://github.com/Speykious/cve-rs), it may be written in a different language with different guarantees, or it may be assembly code with the weakest guarantees possible in the hardware.
+Our goal is to allow the compiler for one language to enforce these properties against all of that other code, irrespective of its source language or its compiler.
+
+Note that this is not about providing *C* memory safety.
+A lot of the work so far is on C and C++, but the same problems occur in safer languages such as Java or Rust.
+For example, here's a snippet from the Rust FFI manual that calls a C function from the `snappy` library:
+
+```rust
+use libc::size_t;
+
+#[link(name = "snappy")]
+extern {
+    fn snappy_max_compressed_length(source_length: size_t) -> size_t;
+}
+
+fn main() {
+    let x = unsafe { snappy_max_compressed_length(100) };
+    println!("max compressed length of a 100 byte buffer: {}", x);
+}
+```
+
+Note that the call from Rust to C requires the `unsafe` keyword.
+This recognises the fact that Rust enforces a lot of useful properties, such as lifetime safety, that C does not and, as soon as you transition into C code, these properties are not automatically enforced.
+As a Rust programmer, you are expected to enforce these properties at the API boundary, but in the general case that is impossible on most hardware.
+For example, there's no Rust wrapper that you can write that prevents a C library (on non-CHERI hardware) from mutating an immutable object, or from capturing a borrowed reference.
+
+But what if the hardware could help?
+What if the Rust compiler could tell some combination of the hardware and the language-agnostic run-time system that a pointer must not be captured, or may be used with load instructions but not store?
+
+## So what is this chariot thing you kids are talking about?
+
```

Comment:
Perhaps start the offspring's response with "CHERIoT.  It's CHERI for the IoT." or such?